frontends/p4/moveDeclarations.cpp

```diff
@@ -21,46 +21,46 @@ namespace P4 {
 const IR::Node *MoveDeclarations::postorder(IR::P4Action *action) {
     if (findContext<IR::P4Control>() == nullptr) {
         // Else let the parent control get these
-        auto body = new IR::IndexedVector<IR::StatOrDecl>();
-        auto m = getMoves();
-        body->insert(body->end(), m->begin(), m->end());
-        body->append(action->body->components);
-        action->body =
-            new IR::BlockStatement(action->body->srcInfo, action->body->annotations, *body);
+        IR::IndexedVector<IR::StatOrDecl> body;
+        const auto &m = getMoves();
+        body.insert(body.end(), m.begin(), m.end());
+        body.append(action->body->components);
+        action->body = new IR::BlockStatement(action->body->srcInfo, action->body->annotations,
+                                              std::move(body));
```

Comment:
Is it actually worth it to use std::move in this case? I understand for the case where you passing along parameters. But here, couldn't the compiler even optimize this? Genuine question.

Reply:
I do not see why compiler would generate something like move ctor here. W/o explicit move it's just plain copy as `body` technically is live after this call, so we need to explicitly "borrow" it. And when copy ctor is generated there is no way back :) copies are elided for certain cases only.

Reply:
My thought was that you could see that the local variable is not used after it was passed in here (with some use-def analysis). Then you could elide the copy. But maybe this is too tricky to implement.

Reply:
This is not what C++ standard says. See https://en.cppreference.com/w/cpp/language/copy_elision

Without this, copy ctor call is emitted. And it is too late for such optimizations as explicit copies and memory allocation is performed that you cannot remove as they are having side effects.

Reply:
I don't think the C++ compiler is permitted to use move instead of copy. The two constructors are allowed to have observably different behavior, even disregarding the state of moved-from object. As far as I know, all optimizations except for copy elision preserve semantics (at least from single-threaded view). And copy elision can only remove copy, it cannot replace it with `move`. The only place where `move` is implicit is in `return foo;` (where in fact this is suitable for copy elision, but in cases it does not happen there will be move). Compiler actually warns about redundant move in that case.

Reply:
In this case the constructors are inline, so the compiler could determine that they don't have observably different behavior, so could be used interchangeably, but it is not required to do so, and it would not surprise me if it did not.

Reply:
> In this case the constructors are inline, so the compiler could determine that they don't have observably different behavior, so could be used interchangeably, but it is not required to do so, and it would not surprise me if it did not.

Yeah I was thinking something along that line but I haven't really seen any discussion on this particular matter online. Mostly for my own curiosity because using std::move in this particular scenario seems onerous.